Phased Strimzi upgrade post (#478)
strimzi/strimzi.github.io: _posts/2025-03-20-phased-strimzi-upgrade-example.md

---
layout: post
title: "Phased upgrades of Strimzi managed Kafka fleets"
date: 2025-03-20
author: tom_cooper
---

> ## TL;DR

Reviewer: I would remove this title, it seems silly and unnecessary. All it does is it provides a simple introduction as all out other blog posts.

When you upgrade an instance of the Strimzi Cluster Operator you will force **all** the Kafka clusters that it manages to roll. 
For large fleets this could be quite disruptive.
It would be nice to be able to decide which Kafka clusters upgrade to the new Strimzi version and when.

This blog provides an example of how phased Strimzi upgrades can be performed.
By using kustomize to create a deployment that allows multiple Strimzi versions to run concurrently and the `STRIMZI_CUSTOM_RESOURCE_SELECTOR` to allow us to specify which Strimzi instance manages which Kafka cluster.

An [example repository](https://github.com/tomncooper/strimzi-phased-upgrade) is provided showing the final setup and providing example scripting and templates for adding new Strimzi versions.

> ## Background

Reviewer: If you move all headers one level down (i.e. `###` instead of `##`, `####` instead of `###`), the actual blog post on the web will look much better and the headers will fit with the rest of the website.

Strimzi handles the [upgrade of Kafka clusters](https://strimzi.io/docs/operators/latest/deploying#proc-upgrade-kafka-kraft-str) for you, making it simple and easy for infrastructure teams to handle these complex operations. 
The [upgrade of the Strimzi operator](https://strimzi.io/docs/operators/latest/deploying#assembly-upgrade-cluster-operator-str) itself is also straight-forward. 
However, when you have a large fleet of Kafka clusters being managed by Strimzi, there are several elements you must keep in mind when performing the upgrade. 

> First, lets go over what a Strimzi upgrade involves:

Reviewer 2: ```suggestion
First, let's go over what a Strimzi upgrade involves:
```

### CRDs

The first step in a Strimzi operator upgrade is the Custom Resource Definitions (CRDs) for the Strimzi Custom Resources (CRs): `Kafka`, `KafkaTopic`, `KafkaUser` etc. 
These are Kubernetes cluster wide resources and need to match the most recent version of Strimzi you have installed on the Kubernetes Cluster. 
They need to match the most recent version as new fields could have been added to the CRDs that the new Strimzi version expects to be there. 

> Strimzi CRDs are backwards compatible (within the same API version), so older Strimzi versions are ok with newer CRDs (they will ignore the new fields) but not the other way around.

Reviewer: Nope, it is not so simple. For most versions, this is valid. But there are always versions when the CRD API versions are changing that will not work in parallel. For example from the past versions, Strimzi 0.21 and 0.23 cannot run in parallel as they will have different CRDs (due to `v1beta2`). Similar situations will likely arise in the future for `v1` API.

While today nobody knows how exactly will the versions look like for `v1` API, the only real guarantee is that you can run in parallel two Strimzi versions that follow each other. (E.g. in the old example you can run 0.21 with 0.22 and 0.22 with 0.23. But not 0.21 or earlier with 0.23 and later.)
The guarantee

Author: This is a good point and why I put the comment in parentheses:

> Strimzi CRDs are backwards compatible (within the same API version)

However, you are right that this needs to be made far more explicit and the consequences discussed in the later text. 

Would it be too strong an assertion to say words to the effect of: 

> provided that the API version is the same, older versions _could_ work with newer CRDs. However the only guaranteed compatibility is between adjacent version, so testing is vital if you plan to have more than two concurrent versions.

I will also add the proviso that if you do have more than 2 concurrent versions and there is an CRD API version change, you will need to upgrade all Kafka clusters to the lower version (the one before the CRD API upgrade) first and then perform the CRD upgrade to the new version (no skipping of the transition Strimzi version).

Reviewer: I think that as long as the API version doesn't change, we do provide backward compatibility. But the problem is that this alone does not provide a future-proof pattern. You essentially do not know exactly when the next API version will arrive. And if you end up having too many different Strimzi versions in use, they might work fine together, but you might find yourself suddenly unable to upgrade as the _next_ version would not work with them. This might be a nasty trap since you might be left out of bugfixes, CVE fixes etc. The ZooKeeper removal makes this even harder as the new versions with `v1` API won't support ZooKeeper anymore, so all Kafka clusters would need to be KRaft-based before you upgrade to v1 API.

With `v1`, we might have hopefully more time to support both versions in parallel as we will not be forced to move quickly by Kubernetes changes. But at this point, we do not know how many versions would be feasible.

So I think the important message here that we need to convey is that:
* These selectors should be fine to use when you want to do the phased upgrade from two adjacent Strimzi versions (which is the core of your blog post, but I suspect many readers will pick out of it what suites them)
* If you want to use this for a longer time to run multiple Strimzi versions in parallel, you need to be aware of the possible consequences. This is also the reason why we were rather silent on this feature (but I'm not saying we should not have this blog post about it - I think it is great to have it, we just need to cover the risks and explain them).

Author: Totally agree, I will make the caveats much stronger and refactor the examples and repo around having just two versions.

Author: Ok I have added a much stronger discussion of the CRD constraints to the start of the Resource Selector Upgrade Section. I have moved the examples to 2 versions of strimzi only and reiterated warnings in the summary and "Things to watch out for" section.

> ### Strimzi Deployment

Reviewer: Should you also mention the cluster roles and cluster role bindings? They will also conflict between parallel operator installations so it should be worth mentioning them.

Author: So this section was supposed to be about what you need to do for a normal Stimzi upgrade and the consequences of that, basically setting up why you might want to have multiple strimzi versions installed.

The ClusterRole/Roles and their bindings are only an issue once you do that and we haven't proposed that yet at this point in the post.

Reviewer: Yet you talk about these things with CRDs right above this, or?

> Once the CRDs are updated, you then need to [upgrade the Strimzi Operator](https://strimzi.io/docs/operators/latest/deploying#assembly-upgrade-cluster-operator-str) its self. 

Reviewer 2: ```suggestion
Once the CRDs are updated, you then need to [upgrade the Strimzi Operator](https://strimzi.io/docs/operators/latest/deploying#assembly-upgrade-cluster-operator-str) itself. 
```

          - name: STRIMZI_KAFKA_CONNECT_IMAGES
            value: |
              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
          - name: STRIMZI_KAFKA_MIRROR_MAKER_IMAGES
            value: |
              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
          - name: STRIMZI_KAFKA_MIRROR_MAKER_2_IMAGES
            value: |
              3.8.0=quay.io/strimzi/kafka:0.45.0-kafka-3.8.0
              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
        ...
```

In the sample above you can see that that Kafka version maps to an image which contains the Strimzi version. 
The container images used by Strimzi include scripts, health endpoint servers and other services which form an API for the container.
A given version of Strimzi expects certain endpoints, scripts etc to be present on the container in order to perform its operations. 

> This is why, even if the Kafka version has not changed, the image must change when you upgrade Strimzi.

Author: That is a good point, but I feel like it is orthogonal to the point of this post. However, we should definitely highlight the custom image Strimzi upgrade issue in the docs (if we don't already).

The result of this is that when you upgrade the Cluster Operator it will, as part of the reconciliation, change the image value on **all** of the Kafka broker pods in all of the Kafka clusters managed by that Strimzi instance on the Kubernetes cluster. 
This means that all of your Kafka fleet (managed by that operator instance) will need to roll. 
Strimzi will do this one broker at time and maintain topic availability (provided your topics have a replication factor greater than 1).
However, rolling every Kafka cluster in your fleet could be quite disruptive.

So what can we do to add more control?

## Phased upgrades

What we need is a way to control which Kafka clusters get upgraded and in what order. 
This is often referred to as a phased upgrade or as an upgrade with _"blast radius"_ control because we can limit the application of the upgrade to a small number of resources (the radius) and contain any potential negative effects (the blast).

There are several possible ways to implement phased upgrades of the Kafka clusters managed by a Strimzi operator:

> ### Custom image per Kafka cluster

Reviewer: This would IMHO not work for most versions and has possibly many different fallouts. You should at least remove this section if not even mention it explicitly as something to absolutely avoid!

Author: Oh I absolutely agree, I only mentioned it as I saw this proposal talked about seriously in the wild and wanted to explain why it was a bad idea.

But removing it seems reasonable, it is not a legitimate option and the post is already quite long.

You can prevent the fleet rolling situation by overriding the [image](https://strimzi.io/docs/operators/latest/configuring#con-common-configuration-images-reference) used by each Kafka cluster. 
You could manually set this to the version used by the current Strimzi Operator version.
As you have manually set this, the changes to the image map environment variables when you upgrade the Strimzi Operator would not change the images of the operands.
So no automatic rolling would occur.
You could then manually switch the image used by each Kafka cluster to the new Strimzi version. 
This would allow you to control which clusters rolled and when.

However, you would then enter a situation where a new operator version (for example 0.45) is talking to a Kafka cluster running the old Strimzi version's container images (eg 0.44). 
As mentioned in the introduction, the Strimzi container images represent an API with certain scripts, endpoints and environment variables expected to be present. 

> Running a newer Strimzi version administering a Kafka cluster using older Strimzi images _could_ lead to issues and is not a recommended or supported deployment model.

Reviewer: Worth adding to this - I forgot it in the initial comment - even with the same image, in most cases the Pods will still roll because there will be other changes as well. E.g. some new setting in the Pod, slightly different certificate subjects, configuration files etc. So even leaving aside the problems, it might still not help to avoid the RU.

Author: Yeah, it is a terrible idea. I have deleted the whole section (in the edited markdown I haven't pushed yet), so we don't accidentally encourage anyone into trying it anyway.

> ### Strimzi per Namespace

Reviewer: It might be worth to emphasize here that only Strimzi versions close to each other can share the CRDs, ClusterRoles etc., so while this is valid way, it requires a lot of self-control to make sure the versions do not diverge too much.

However, if you have multiple Kafka clusters in a namespace you would still roll all of them automatically. 
More seriously, as a single Strimzi instance is capable of running multiple Kafka clusters, this is a very inefficient deployment model with potentially many Strimzi instances running idle. 

The Cluster Operators would also all share the cluster-wide CRDs and so care and time would need to be taken to update these and upgrade all the Operators in a timely fashion.

### Pause Reconciliation

If you have a Strimzi Cluster Operator managing multiple Kafka clusters, either across a whole Kubernetes cluster or in a namespace, then another option is to temporarily stop the Operator from managing a given Kafka cluster (or other operand) until the upgrade is complete.

To do this you would need to:
1) Apply the [pause reconciliation annotation](https://strimzi.io/docs/operators/latest/deploying#proc-pausing-reconciliation-str) to all Kafka CRs (and other operands like Kafka Connect deployments)
2) Upgrade the CRDs and the Strimzi Operator Deployment to the new version.

> 3) In the order you want, remove the annotation from each Kafka CR. The operator will resume reconciliation, update the image reference and roll the Kafka cluster.

Reviewer: This does not seem to render. Do you need to use `1.`, `2.` etc.? https://deploy-preview-478--strimzi.netlify.app/blog/2025/03/20/phased-strimzi-upgrade-example/

> 3) In the order you want, remove the annotation from each Kafka CR. The operator will resume reconciliation, update the image reference and roll the Kafka cluster.

Reviewer: You resolved the original comment, but it is still wrong.

Author: Sorry, haven't actually pushed up the edited post markdown file yet. I should have done that before resolving.

The two Strimzi deployments would coexist happily (provided you have setup independent RBAC -- more on this later), managing completely separate sets of CRs.

This features allows us to create a more robust upgrade process for the Strimzi Operator and its operands.

## Resource Selector based upgrade

Using the Custom Resource Selector (CRS) allows us to have fine grained control over which operands are upgraded to new Strimzi versions and when. 
However, in order to use this process we need to deploy multiple versions of Strimzi concurrently.

### Concurrent Strimzi Deployments

The standard Strimzi deployment includes service accounts and RBAC resources (`Role`s, `ClusterRole`s and their associated bindings) needed for Strimzi to do its job. 
By default these are all named `strimzi-cluster-operator` or similar.
If we try and deploy multiple operator instances in the same namespace, or over multiple namespaces but needing to share `ClusterRole`s and bindings, things will get messy.
Also, different Strimzi versions may need different permissions and so we need to make sure independent copies of these RBAC resources are deployed for each Strimzi version.

> #### Deploying Strimzi with Kustomize

Reviewer: If you are happy with this, no need to change. But I wonder if it would be easier to just use different namespaces for each operator version?

- 021-ClusterRole-strimzi-cluster-operator-role.yaml
- 022-ClusterRole-strimzi-cluster-operator-role.yaml
- 022-RoleBinding-strimzi-cluster-operator.yaml
- 023-ClusterRole-strimzi-cluster-operator-role.yaml
- 023-RoleBinding-strimzi-cluster-operator.yaml
- 030-ClusterRoleBinding-strimzi-cluster-operator-kafka-broker-delegation.yaml
- 030-ClusterRole-strimzi-kafka-broker.yaml
- 031-ClusterRole-strimzi-entity-operator.yaml
- 031-RoleBinding-strimzi-cluster-operator-entity-operator-delegation.yaml
- 033-ClusterRoleBinding-strimzi-cluster-operator-kafka-client-delegation.yaml
- 033-ClusterRole-strimzi-kafka-client.yaml
- 050-ConfigMap-strimzi-cluster-operator.yaml
- 060-Deployment-strimzi-cluster-operator.yaml
``` 

`install-files/crds/kustomization.yaml`
```yaml
resources:
- 040-Crd-kafka.yaml
- 041-Crd-kafkaconnect.yaml
- 042-Crd-strimzipodset.yaml
- 043-Crd-kafkatopic.yaml
- 044-Crd-kafkauser.yaml
- 045-Crd-kafkamirrormaker.yaml
- 046-Crd-kafkabridge.yaml
- 047-Crd-kafkaconnector.yaml
- 048-Crd-kafkamirrormaker2.yaml
- 049-Crd-kafkarebalance.yaml
- 04A-Crd-kafkanodepool.yaml
```

`install-files/crds/kustomization.yaml`
```yaml
resources:
- cluster-operator-0.43.0
- crds

> ```

Reviewer 2: this long list of folders + files could easily confuse readers to understand what they are looking at. Maybe a short description for each would help?

Author: I added more descriptive captions for each listing, hopefully that makes it clearer what each relates too.

It also updates any references to those renamed resources.
For example this will change the name of the Strimzi configuration `ConfigMap` resource and also the reference to it in the Strimzi Cluster Operator `Deployment` resource.

To make this change we simply add one line to the `kustomization.yaml` in each Strimzi version directory:

`install-files/cluster-operator-0.43.0/kustomization.yaml`
```yaml
resources:
- 010-ServiceAccount-strimzi-cluster-operator.yaml
...
- 060-Deployment-strimzi-cluster-operator.yaml

nameSuffix: '-0-43-0'
```

That's it. 
You can see how it changes the names by building the kustomization with:
```shell
kubectl kustomize install-files
```

We can now add a newer Strimzi version by downloading the install files for that version, moving the CRDs to the separate `crds` folder (replacing the older ones) and then adding a `kustomization.yaml` to that version's folder with a different `nameSuffix`:

`install-files/cluster-operator-0.44.0/kustomization.yaml`
```yaml
resources:
- 010-ServiceAccount-strimzi-cluster-operator.yaml
...
- 060-Deployment-strimzi-cluster-operator.yaml

nameSuffix: '-0-44-0'
```

#### Patching Environment Variables

> Sadly, the `nameSuffix` is not sufficient on it own. 

Reviewer 2: ```suggestion
Sadly, the `nameSuffix` is not sufficient on its own. 
```

metadata:
  name: strimzi-cluster-operator
spec:
  template:
    spec:
      containers:
        - name: strimzi-cluster-operator
          env:
            - name: STRIMZI_NAMESPACE
              value: "*" 
              valueFrom: null
            - name: STRIMZI_CUSTOM_RESOURCE_SELECTOR
              value: strimzi-resource-selector=strimzi-0-43-0
            - name: STRIMZI_LEADER_ELECTION_LEASE_NAME
              value: "strimzi-cluster-operator-0-43-0"
```

[The documentation](https://strimzi.io/docs/operators/latest/deploying#deploying-cluster-operator-to-watch-whole-cluster-str) then says we need to create several additional (Cluster)Roles and bindings. 
We can add these to the base install resources by using the `--dry-run=client` and `-o yaml` flags for the `kubectle create` command to output the resource files to yaml format and then send that to a file.
For example:

```shell
kubectl create --dry-run=client -o yaml \
  clusterrolebinding strimzi-cluster-operator-namespaced \
  --clusterrole=strimzi-cluster-operator-namespaced \
  --serviceaccount strimzi:strimzi-cluster-operator \
  > install-files/cluster-operator-0.43.0/strimzi-cluster-operator-namespaced.yaml
```

> We can then add these additional resources to the `resources` list in the respective `kustomization.yaml` files and they will be modified by the same `nameSuffix` and `patch-role-reference.yaml` patches as the other RBAC resources.

Reviewer: Doesn't this just create a new installation file while keeping the old RoleBinding in place as well?

Author: Yes, but with this setup all the RBAC is completely seperate, different names pointing at different roles and different service accounts.

│  │  ├── patch-deployment.yaml
│  │  └── patch-role-references.yaml
│  ├── 010-ServiceAccount-strimzi-cluster-operator.yaml
│  ├── ...
│  ├── 060-Deployment-strimzi-cluster-operator.yaml
│  ├── kustomization.yaml
├── crds
│  ├── 04A-Crd-kafkanodepool.yaml
│  ├── 040-Crd-kafka.yaml
│  ├── 041-Crd-kafkaconnect.yaml
│  ├── 042-Crd-strimzipodset.yaml
│  ├── 043-Crd-kafkatopic.yaml
│  ├── 044-Crd-kafkauser.yaml
│  ├── 045-Crd-kafkamirrormaker.yaml
│  ├── 046-Crd-kafkabridge.yaml
│  ├── 047-Crd-kafkaconnector.yaml
│  ├── 048-Crd-kafkamirrormaker2.yaml
│  ├── 049-Crd-kafkarebalance.yaml
│  └── kustomization.yaml
└── kustomization.yaml
```

To install the entire Operator Set you simply run:

```shell
kubectl -n strimzi apply -k install-files
```

Note that the namespace you install the Operator set into needs to match the one you have used in the `patch-role-references.yaml`.

A full example of the Operator Set above can be seen in the [strimzi-phased-upgrade](https://github.com/tomncooper/strimzi-phased-upgrade) repository.
That repository also contains scripting to automate the addition of new Strimzi versions and templates for the generation of patch and kustomization files. 

> ### The Phased Upgrade Process

Reviewer: As I said above, I think you should maybe operate with just two versions as three might become too much.

We think we know what the issue is and how to fix it, but as this is a critical cluster we need to roll back and make sure all the dependant services are running correctly. 
In this setup the roll back is a simple label change:

```shell
kubectl -n kafka label --overwrite kafka kafka-a strimzi-resource-selector=strimzi-0-43-0
```

```mermaid
flowchart TD
    S_043(Strimzi 0.43) --> K_A(Kafka A)
    S_044(Strimzi 0.44) --> K_B(Kafka B)
	S_044 --> K_C(Kafka C)
	S_045(Strimzi 0.45)
```

We fix the configuration issue and move Kafka cluster A over to the 0.44 operator.
We then move cluster C onto the latest 0.45 operator, ready to upgrade the rest of the fleet.

```mermaid
flowchart TD
    S_043(Strimzi 0.43)
	style S_043 stroke:#f66,stroke-width:2px
    S_044(Strimzi 0.44) --> K_A(Kafka A)
	S_044 --> K_B(Kafka B)
	S_045(Strimzi 0.45) --> K_C(Kafka C)
```

At this point the Strimzi 0.43 operator is not managing any Kafka clusters and all the clusters being managed by the 0.44 operator have been stable for sufficient period that we are confident we don't need to roll back. 
We can now remove the Strimzi 0.43 operator deployment and its associated RBAC resources.

> ### Things to watch out for

Reviewer: You should add here also a warning about not removing the CRDs or ClusterRoles by mistake when deleting the old version. Especially with the CRDs, it would wipe all operands. This is the main issue of this approach.

Reviewer 2: Yeah. Maybe it should also be mentioned at the beginning when dealing with pros and cons.

Author: It is definatly something to be mindful of so I will add it here. I'm not sure about adding at the beginning as it is a risk with a normal Strimzi deployment as well.

Author: I added this caveat to the summary section at the top of the post.
